framework/include/reporters/ReporterData.h

```diff
@@ -0,0 +1,222 @@
+   * scatter and broadcast actions via the special call on the storage helper object. This
+   * is a replacement for that method that leverages the RepoorterContext objects to perform
+   * value specific actions.
+   *
+   * See FEProblemBase::joinAndFinalize
+   */
+  void finalize(const std::string & object_name);
+
+  /**
+   * At the end of a timestep this method is called to copy the values back in time in preparation
+   * for the next timestep.
+   *
+   * See FEProblemBase::advanceState
+   */
+  void copyValuesBack();
+
+  /**
+   * Copies the current value to all the desired old values and shrinks the old value vector to
+   * the correct size based on the requested values.
+   *
+   * See InitReporterAction
+   */
+  void init();
+
+private:
+  /// For accessing the restart/recover system, which is where Reporter values are stored
+  MooseApp & _app;
+
+  /**
+   * Helper object that for creating the necessary RestartableData for Reporter values.
+   * @tparam T The desired C++ type for the Reporter value
+   * @param reporter_name Object/data name for the Reporter value
+   * @param declare Flag indicating if the ReporterValue is being declared or read. This flag
+   *                is passed to the existing MooseApp restart/recover system that errors if a
+   *                value is declared multiple times.
+   */
+  template <typename T>
+  ReporterState<T> & getReporterStateHelper(const ReporterName & reporter_name, bool declare);
+
+  /**
+   * Helper object for declaring Reporter values.
+   *
+   * The ReporterContext objects allow for custom handling of data (e.g., broadcasting the value),
+   * these context objects use the context system that exists on the MooseApp restart/recover
+   * system. However, the get/declare methods can be called in any order thus an the underlying
+   * RestartableData object is often created by the get method before it is declared. The get
+   * methods should not have any knowledge of the ReporterContext object. As such the context object
+   * must be applied to the RestartableData object after creation. This helper does just that.
+   */
+  template <typename T, template <typename> class S>
+  ReporterState<T> & declareReporterValueHelper(const ReporterName & reporter_name);
+
+  /// Convenience for accessing all the ReporterValues for a given Reporter object, the
+  /// map key is the ReporterObject name. The data pointers are used by the init, finalize, and
+  /// copyValuesBack methods. Those functions could operate using the MooseApp interface, but
+  /// this is much more convient.
+  std::unordered_map<std::string, std::set<RestartableDataValue *>> _data_ptrs;
+
+  /// The ReporterContext objects are "attached" to the RestartableDataValue objects via a void
+  /// pointer. Therefore, the actual object must be owned and managed by something, this is the
+  /// something.
+  std::set<std::unique_ptr<ReporterContextBase>> _context_ptrs;
```

Comment:
This is a WIP. I would like to see these go away and just lookup stuff in the RestartableDataMap object in MooseApp, I just haven't come up with a good way to do that yet. Another option is to just use the ReporterContext objects, to access data because as is they reference the RestartableData anyway.

Reply:
I improved this in the lastest commit